web/src/ui.rs

```diff
@@ -116,6 +120,91 @@ impl FileDialogResult for WebFileDialogResult {
     fn refresh(&mut self) {}
 }
 
+pub struct WebDownloadFileData {
+    file_name: String,
+    contents: Vec<u8>,
+    downloaded: bool,
+}
+
+pub struct WebDownloadFileDialogResult {
+    data: Mutex<WebDownloadFileData>,
+}
+
+impl WebDownloadFileDialogResult {
+    fn new(file_name: String) -> WebDownloadFileDialogResult {
+        Self {
+            data: Mutex::new(WebDownloadFileData {
+                file_name,
+                contents: Vec::new(),
+                downloaded: false,
+            }),
+        }
+    }
+
+    fn data(&self) -> MutexGuard<'_, WebDownloadFileData> {
+        self.data.lock().expect("non-poisoned data")
+    }
+
+    fn download(mut data: MutexGuard<WebDownloadFileData>) -> Option<()> {
+        let array = Uint8Array::from(&data.contents[..]);
+        let blob = Blob::new_with_u8_array_sequence(&array).ok()?;
+        let window = web_sys::window()?;
+        let document = window.document()?;
+        let a = document.create_element("a").ok()?;
+
+        let url = JsUrl::create_object_url_with_blob(&blob).ok()?;
+        a.set_attribute("href", &url).ok()?;
+        a.set_attribute("download", &data.file_name).ok()?;
+        a.dyn_into::<HtmlElement>().ok()?.click();
+        data.downloaded = true;
+        Some(())
+    }
+}
+
+impl FileDialogResult for WebDownloadFileDialogResult {
+    fn is_cancelled(&self) -> bool {
+        self.data().downloaded
+    }
+
+    fn creation_time(&self) -> Option<DateTime<Utc>> {
+        None
+    }
+
+    fn modification_time(&self) -> Option<DateTime<Utc>> {
+        None
+    }
+
+    fn file_name(&self) -> Option<String> {
+        Some(self.data().file_name.clone())
+    }
+
+    fn size(&self) -> Option<u64> {
+        Some(self.data().contents.len() as u64)
+    }
+
+    fn file_type(&self) -> Option<String> {
+        None
+    }
+
+    fn creator(&self) -> Option<String> {
+        None
+    }
+
+    fn contents(&self) -> &[u8] {
+        tracing::warn!("Reading downloaded file not implemented");
+        &[]
```

Comment:
I actually thought that using `load()` after everything from `save` is complete would allow access to data, but I currently can't reproduce that anymore. Something to look into.

Reply:
That's definitely on my TODO list

Reply:
TBH I also had some trouble testing it, so I decided to basically make every `FileDialogResult` implementation behave the same way. That was possible by joining `write` and `refresh` to `write_and_refresh`. This way the SWF tests also verify whether this behavior is correct.

Now I'm only wondering whether it's covered by SWF tests sufficiently :thinking:

Reply:
Okay, I really tried to break it (call load after save, or something), but it seems that FP always throws error 2037 when the order is not right, it's much more stricter than Ruffle currently (we should probably model it as a state enum and disallow some transitions).

The only discrepancy I found was related to reporting the file size during saving (it was even commented out in tests), but that's a problem not related to this PR (and it affects all platforms).

Reply:
I had actually started experimenting with type states, but I wasn't really in the mood of rewriting a lot of the code after it already took weeks to land the initial version. I think it's not a bad idea, but testing all the FP edge cases will get annoying.

More test are always appreciated.